### DISCUSSION TITLE
Which tutorial?

### DISCUSSION THREAD

**Discussion by dstromberg** [sqlalchemy/sqlalchemy#9775] [Usage Questions]
Hi.

I'm thinking I should dig into SQLAlchemy to understand why some of my ORM queries are having problems.  It seems like SA's laziness is making this a challenge that will require some dedication to get a good handle on.

We until recently were using SA 1.2.5 on Python 2.7, but recently switched to SA 1.4.36 on Python 3.10 and we're not yet using the "future" mode of 1.4.x.

I looked at https://docs.sqlalchemy.org/en/14/tutorial/ and it looks interesting, but does it only cover future mode?

I suspect I'm best off learning 1.2 for now - though hopefully that suspicion is unfounded.

I'm working in a large codebase that was written by someone else, and that uses SA quite a lot.

What SA tutorial should I study?

Thanks!

**Comment by CaselIT** [comment]
Hi,

The 1.4 and 2.0 tutorial were largely re-written to use the new syntax. For a greenfield project the new 2.0 tutorials are the ones to check.

But since in your case you have to maintain a large codebase that still heavely features the older style it probably makes sense first getting comfortable with it by looking mainly at the 1.3 docs, then check the changes introduced in 1.4 and v2 (start by looking at the migration notes) so that new code could use the new syntax. Note that there are no plans of ever removing `query`, but the new style is nicer to use

**Reply by dstromberg** [reply]
How about https://docs.sqlalchemy.org/en/13/orm/tutorial.html ?

Thanks!

**Reply by CaselIT** [reply]
For understanding the 1.3 code base that's the one to look.
But for new code the v2 one should be preferred

**Reply by dstromberg** [reply]
I may try:
https://web.archive.org/web/20201124155702/https://docs.sqlalchemy.org/en/12/orm/tutorial.html

It's 1.2.  Or are 1.2 and 1.3 so similar that I don't need archive.org?

**Reply by CaselIT** [reply]
From memory 1.2 and 1.3 syntax is very similar, so it should not be a problem looking al the 1.3 version.
